server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -1801,16 +1868,57 @@ public int deletePendingSegments(String dataSource)
     );
   }
 
+  private boolean shouldPersistSchema(SegmentSchemaMapping segmentSchemaMapping)
+  {
+    return schemaPersistEnabled
+           && segmentSchemaMapping != null
+           && segmentSchemaMapping.isNonEmpty();
+  }
+
+  private void persistSchema(
+      final Handle handle,
+      final Set<DataSegment> segments,
+      final SegmentSchemaMapping segmentSchemaMapping
+  ) throws JsonProcessingException
+  {
+    if (segmentSchemaMapping.getSchemaVersion() != CentralizedDatasourceSchemaConfig.SCHEMA_VERSION) {
+      log.error(
+          "Schema version [%d] doesn't match the current version [%d]. Not persisting this schema [%s]. "
+          + "Schema for this segment will be populated by the schema backfill job in Coordinator.",
+          segmentSchemaMapping.getSchemaVersion(),
```

Comment:
This check should be outside the transaction. Lets create a follow up patch for that.